### DISCUSSION TITLE
alm CV.in process is very slow, are there any strategies to speed it up?

### DISCUSSION THREAD

**Discussion by chunweizhu** [ttadano/alamode#182] [Q&A]
Dear developers
    In the cross validation process, to determine the optimal amplitude of regularization ($\alpha$) is very slow. Is there any acceleration strategy?
Best regards

**Comment by ttadano** [comment]
How slow is it? Did you use the default value of [`STOP_CRITERION`](https://alamode.readthedocs.io/en/latest/almdir/inputalm.html#alm-stop-criterion)?

**Reply by chunweizhu** [reply]
Hi @ttadano 
Yes, I'm using the default STOP_CRITERION value, and for an 80-atom structure, 202 data set, it takes two or three days.

**Reply by ttadano** [reply]
That is actually slower than I expected, but the actual computational time depends on the number of free parameters. 

If you believe the low performance of the cross validation is due to the implementation in ALAMODE, I suggest that you try the external implimentations, such as scikit-learn. 

For more details, please have a look at https://github.com/ttadano/ALM/tree/develop/example/lasso.

**Reply by chunweizhu** [reply]
Hi @ttadano  
Thank you for your advice. 😊
The following is the log file during runtime. The number of free parameters is `28434`. Is this too much?

~~~md
 CONSTRAINT
 ==========

  ICONST = 1: Constraints for translational invariance
              will be considered.

  FC2XML is given : Harmonic force constants will be 
                    fixed to the values given in harm.xml

  Generating constraints for translational invariance ...
   HARMONIC ... done.
    ANHARM3 ... done.
    ANHARM4 ... done.
  Finished !

  Number of constraints [T-inv, R-inv (self), R-inv (cross)]:
   HARMONIC     49    0    0
    ANHARM3   3248    0    0
    ANHARM4   5679    0    0

  Constraints of T-inv and R-inv (self) are merged.
  If there are redundant constraints, they are removed in this process.

  Number of inequivalent constraints (self, cross) : 
   HARMONIC     49    0
    ANHARM3   3248    0
    ANHARM4   5679    0

  ICONST >= 10 : Constraints will be considered algebraically.

  Number of free HARMONIC FCs : 0
  Number of free  ANHARM3 FCs : 24753
  Number of free  ANHARM4 FCs : 3681

  Time Elapsed: 12.8769 sec.

 -------------------------------------------------------------------

 OPTIMIZATION
 ============

  LMODEL = elastic-net

  Training data file (DFSET) : DFSET_aimd_randoms@200

  NSTART = 1; NEND = 202
  202 entries will be used for training.

  Total Number of Parameters : 39261
  Total Number of Free Parameters : 28434

 Elastic-net  cross-validation with the following parameters:
   L1_RATIO = 1
   CV =               4
   CV_MINALPHA =           1e-08  CV_MAXALPHA =            0.01
   CV_NALPHA =    30
   CONV_TOL =           1e-08
   MAXITER = 10000
   STOP_CRITERION =     5
   ENET_DNORM =               1

  STANDARDIZE = 1 : Standardization will be performed for matrix A and vector b.
                    The ENET_DNORM-tag will be neglected.

  CV > 0: Automatic CV mode.

  Start 4-fold CV with 202 Datasets


  SET :   1
  Recommended CV_MAXALPHA = 0.00242291
   ......
~~~

**Reply by ttadano** [reply]
Yes, this is a large problem. Can you make the cutoff radii smaller, check the validation score, and gradually increase the radii if necessary?

**Reply by chunweizhu** [reply]
Hi @ttadano, 
I'm going to do a test like this, Thank you for your patient answer.😊
Best regards

**Reply by chunweizhu** [reply]
Hi @ttadano 
I'm sorry, I would like to ask you another small question. 
When using the extract.py script：
~~~python
$ python ${ALAMODE_ROOT}/tools/extract.py --VASP SPOSCAR --offset vasprun0.xml vasprun_harm*.xml > DFSET_harmonic
~~~
To obtain the correct DFSET file, does the atomic order in POSCAR (used to obtain vasprun_harm*.xml) and the atomic order in the SPOSCAR file have to maintain a one-to-one correspondence?

----

SPOSCAR                                                 
.........                                                         
                                                                
$x_1$  $y_1$ $z_1$     atom1                   
$x_2$  $y_2$ $z_2$     atom2                   
$x_3$  $y_3$ $z_3$     atom3                   
$x_4$  $y_4$ $z_4$     atom4 
**if the SPOSCAR file atomic order is `atom1 atom2 atom3 atom4`**   
              
----

POSCAR                                              
.........                                                         

$x_2$  $y_2$ $z_2$     atom2                                                                  
$x_1$  $y_1$ $z_1$     atom1                              
$x_3$  $y_3$ $z_3$     atom3                   
$x_4$  $y_4$ $z_4$     atom4  
**So the POSCAR file atomic order is `atom2 atom1 atom3 atom4`**  that can not obtain the correct DFSET file?

**Reply by ttadano** [reply]
Yes, the order of atoms in the supercell is important and must be consistent throughout the calculations.

**Reply by chunweizhu** [reply]
Hi @ttadano 
Thank you for your answer it helps me a lot.😊
Best regards
